src/charm.py

```diff
@@ -227,6 +245,64 @@ def _on_database_storage_detaching(self, _) -> None:
         # Inform other hooks of current status
         self._peers.data[self.unit]["unit-status"] = "removing"
 
+    def _on_update_status(self, _) -> None:
+        """Handle update status.
+
+        Takes care of workload health checks and.
+        """
+        if not self.cluster_initialized or not self.unit_peer_data.get("member-role"):
+            # health checks only after cluster and member are initialised
+            return
+
+        # retrieve and persist state for every unit
+        try:
+            state, role = self._mysql.get_member_state()
+            self.unit_peer_data["member-role"] = role
+            self.unit_peer_data["member-state"] = state
+        except MySQLGetMemberStateError:
+            if self.unit_peer_data.get("member-state") == "waiting":
+                # avoid changing status while in initialisation
+                return
+            role = self.unit_peer_data["member-role"] = "unknown"
+            state = self.unit_peer_data["member-state"] = "unreachable"
+        logger.info(f"Unit workload member-state is {state} with member-role {role}")
+
+        # set unit status based on member-{state,role}
+        self.unit.status = (
+            ActiveStatus(f"Unit is ready: Mode: {'RW' if role == 'primary' else 'RO'}")
+            if state == "online"
+            else MaintenanceStatus(state)
+        )
+
+        if state == "recovering":
+            # server is in the process of becoming an active member
+            logger.info("Instance is being recovered")
+            return
+
+        if state == "offline":
+            # Group Replication is active but the member does not belong to any group
+            all_states = {
+                self._peers.data[unit].get("member-state", "unknown") for unit in self._peers.units
+            }
+
+            if all_states == {"offline"}:
```

Comment:
there are 2 consecutive `if` statements, which can be combined into one to reduce the cyclomatic complexity of the handler:
```
if all_status == {"offline"} and self.unit.is_leader():
```